File: algorithms/dfs/c/main_dfs.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include "dfs.h"  // Include dfs.h from the same folder


#define MAX_LINE_LENGTH 1024
#define MAX_NODES 100

// Function to load a graph from a dataset file
Graph loadGraphFromDataset(const char *datasetPath) {
    FILE *file = fopen(datasetPath, "r");
    if (file == NULL) {
        perror("Failed to open dataset file");
        exit(EXIT_FAILURE);
    }

    Graph graph;
    initGraph(&graph, MAX_NODES);

    char line[MAX_LINE_LENGTH];
    while (fgets(line, sizeof(line), file)) {
        int from;
        char neighbors[MAX_LINE_LENGTH];
        if (sscanf(line, "%d,%s", &from, neighbors) == 2) {
            char *neighbor = strtok(neighbors, ",");
            while (neighbor != NULL) {
                int to = atoi(neighbor);
                addEdge(&graph, from, to);
                neighbor = strtok(NULL, ",");
            }
        }
    }

    fclose(file);
    return graph;
}

// Main function
int main() {
    const char *datasetPath = "datasets/graph_datasets/gigantic_graph_dataset.csv";
    int startNode = 0;  // Starting node for DFS
    int visited[MAX_NODES] = {0};

    // Load the graph dataset
    printf("Loading graph dataset from: %s\n", datasetPath);
    Graph graph = loadGraphFromDataset(datasetPath);

    // Log start time
    clock_t startTime = clock();

    // Perform DFS traversal
    printf("DFS Traversal starting from node %d:\n", startNode);
    dfs(&graph, startNode, visited);

    // Log end time and calculate elapsed time
    clock_t endTime = clock();
    double timeElapsed = ((double)(endTime - startTime)) / CLOCKS_PER_SEC;
    printf("\nTime taken for DFS: %.6f seconds\n", timeElapsed);

    return 0;
}

File: algorithms/dfs/python/main_dfs.py
# algorithms/dfs/python/main_dfs.py

import importlib
import pandas as pd
import time

def load_algorithm(algorithm_path, algorithm_name, function_name):
    module = importlib.import_module(f"{algorithm_path}.{algorithm_name}")
    return getattr(module, function_name)

def load_graph_dataset(dataset_path):
    df = pd.read_csv(dataset_path, header=None, names=["Node", "Neighbors"], quotechar='"')
    graph = {}
    
    # Step 1: Populate the graph from the dataset
    for _, row in df.iterrows():
        node = row["Node"]
        neighbors = row["Neighbors"].split(",") if pd.notna(row["Neighbors"]) else []
        graph[node] = neighbors

    # Step 2: Collect all missing nodes
    missing_nodes = set()
    for neighbors in graph.values():
        for neighbor in neighbors:
            if neighbor not in graph:
                missing_nodes.add(neighbor)

    # Step 3: Add missing nodes to the graph with empty neighbors
    for missing_node in missing_nodes:
        graph[missing_node] = []

    return graph


def main():
    algorithm_path = 'algorithms.dfs.python'
    algorithm_name = 'dfs'
    function_name = 'dfs'
    dataset_path = 'datasets/graph_datasets/gigantic_graph_dataset.csv'

    dfs_function = load_algorithm(algorithm_path, algorithm_name, function_name)
    graph = load_graph_dataset(dataset_path)
    start_node = 'N1'

    start_time = time.time()
    visited_nodes = dfs_function(graph, start_node)
    end_time = time.time()

    print(f"DFS traversal starting from {start_node}: {visited_nodes}")
    print(f"Time taken: {end_time - start_time:.6f} seconds")

if __name__ == "__main__":
    main()

# python -m algorithms.dfs.main_dfs
